content/struct-classes.md

```diff
@@ -0,0 +1,627 @@
+changed because the proposal does not change the existing language features, it only
+adds a new keyword, `struct`.
+
+### Open questions
+
+#### Choice of keyword `struct`
+
+The keyword `struct` stands for “structural”, which relates directly to the semantics
+of `struct class` definitions: class definitions that automatically implement
+structural equality (as opposed to the default object reference equality).
+
+However, `struct` has historically also been used in C and C++ to define “data structures”,
+with no relation with structural equality. The keyword `struct` is also used in Swift and
+Rust with a similar meaning (“data structure”).
+
+Other languages implementing a similar feature use the following keywords:
+- `record` (Java), a restricted variant of `struct class` (Java records can’t
+  extend parent classes, can’t be extended, and can’t have mutable fields --- even
+  private ones),
+- `data class` (Kotlin), which are equivalent to `case classes` in Scala and have
+  the same issues regarding binary compatibility.
+
+We think that using the keyword `record` or `data` instead of `struct` would bring
+confusion because the feature these keywords relate to in other programming languages
+is closer to Scala `case classes` than `struct classes`. On the other hand, we agree
+that in other programming languages the keyword `struct` may have an established
+meaning that is different from the one we give here, so we are open to using a
+different keyword, such as `structural`.
+
+## Alternatives
+
+### Automatic generation of transformation methods
+
+Besides the two solutions shown in the Motivation section (based on regular classes
+or case classes), we also considered a more powerful variant of `struct class` that
+would also automatically generate the transformation methods (`withName`, `withAge`, 
+etc. in the example). However, this solution is more complex to specify, and it also
+has some annoying drawbacks. In particular, it does not work well with all the names,
+especially with symbolic names (e.g., `+`, `%`, etc.), or names that contain
+acronyms (e.g. consider a field `httpHeaders`, should the transformation method
+be named `withHttpHeaders` or `withHTTPHeaders`?). Furthermore, it would be extremely
+complicated to specify rules to handle "smarter" transformation methods, such as
+`withoutEmail` in our example, to specifically handle fields whose type is `Option`,
+or a collection.
+
+### Leverage `Product`
+
+We could introduce a new trait, say `Structural`, to the standard library,
+which would implement the methods `toString`, `hashCode`, and `equals` in 
+terms of methods provided by the trait `Product`:
+
+~~~ scala
+trait Structural extends Product:
+  override def hashCode: Int =
+    productIterator.foldLeft(17)((hash, field) => 37 * (hash + field.##))
+  override def equals(other: Any): Boolean =
+    other match
+      case that: Structural if that.canEqual(this) =>
```

Comment:
Wouldn't it be better to turn it around and have
```scala
... if this.canEqual(that) =>
```
So that I can potentially override `canEqual` in the `Structural` sub-class in those rare cases where I want to such thing?

Reply:
I’ve followed the pattern described here: https://www.artima.com/pins1ed/object-equality.html

Reply:
I see. But still. Don't you see utility in a sub-class of `Structural` to decide to be `equals`able with other classes, for example case classes?
```scala
class UserDto(name: String, age: Int)

class User(val name: String, val age: Int, val email: Option[String]) extends Structural:
  private def copy(name: String = this.name, age: Int = this.age, email: Option[String] = this.email): User =
    User(name, age, email)
  def with...
  def canEqual(other: Any): Boolean = other match:
    case _: User => true
    case _: UserDto => true
    case _ => false

User("John", 21, None) == UserDto("John", 21)  // returns `true`
User("John", 21, Some("john@email.com")) == UserDto("John", 21)  // also returns `true`
```